### PR TITLE
daemon: support config url

### PR DESCRIPTION
**Pull request by FeelyChau**
#200 
Support config-url.
You can run as:
```shell
$ pipcook run http://host/config.json
```

### CODE REVIEW

**Comment by yorkie** [comment]
@FeelyChau Thanks for this contribution, and just have a double check before you submitting a pull request next time:

- run `npm run lint` locally to make sure the lint pass.
- it's recommended that using `fix/gh-{id}` as the branch name if you are going to fix issue on GitHub.

File: packages/cli/src/utils.ts [hunk 1]
```diff
@@ -24,3 +24,6 @@ export function tail(id: string, name: string): ChildProcess {
     }
   );
 }
+export function isUrl(url: string) {
+  return /^https?:\/\/.+/.test(url)
```
Comment (yorkie) [inline packages/cli/src/utils.ts:28]: How about Node.js `url.parse()`?

File: packages/daemon/src/runner/helper.ts [hunk 2]
```diff
@@ -13,8 +14,13 @@ import {
 
 const { PLUGINS, PIPCOOK_LOGS } = constants;
 
-export async function parseConfig(configPath: string, generateId = true): Promise<PipelineDB> {
-  const configJson: RunConfigI = await fs.readJson(configPath);
+export async function parseConfig(config: string, generateId = true): Promise<PipelineDB> {
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:17]: The function signature should not be changed :)

File: packages/daemon/src/runner/helper.ts [hunk 3]
```diff
@@ -13,8 +14,13 @@ import {
...
+  let configJson: RunConfigI = null;
+  if (/^https?:\/\/.+/.test(config)) {
+    configJson = JSON.parse(await request(config));
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:20]: Maybe we need a `loadJson()` at utils?

File: packages/cli/src/actions/start.ts [hunk 4]
```diff
@@ -15,10 +15,12 @@ const start: StartHandler = async (filename: string, opts: any) => {
     spinner.fail('Please specify the config path');
     return process.exit(1);
   }
-  filename = path.isAbsolute(filename) ? filename : path.join(process.cwd(), filename);
-  if (!existsSync(filename)) {
-    spinner.fail(`${filename} not exists`);
-    return process.exit(1);
+  if (!isUrl(filename)) {
```
Comment (yorkie) [inline packages/cli/src/actions/start.ts:18]: Why do we add this check?

File: packages/cli/src/utils.ts (hunk 1 above)
Comment (FeelyChau) [reply packages/cli/src/utils.ts:28]: This PR supports http and https only, `url.parse()` is out of the protocols.

File: packages/cli/src/actions/start.ts (hunk 4 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts:18]: If not, the file url will be joined the file path prefix, we should skip this step if it's a url.

Comment (yorkie) [reply packages/cli/src/actions/start.ts:18]: How about moving the existence check to daemon? So here is to do convert filename to file-protocol URI, and the daemon API might handle the URI only.

File: packages/cli/src/utils.ts (hunk 1 above)
Comment (yorkie) [reply packages/cli/src/utils.ts]: `url` can handle http/https, try `urlobj.protocol`.

File: packages/daemon/src/runner/helper.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts:18]: But the meaning of `configPath` has actually changed.

File: packages/daemon/src/runner/helper.ts (hunk 3 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts]: Use `url.parse()` to get the url protocol.

File: packages/cli/src/actions/start.ts (hunk 4 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts]: > How about moving the existence check to daemon? So here is to do convert filename to file-protocol URI, and the daemon API might handle the URI only.

It looks more reasonable.

File: packages/daemon/package.json [hunk 5]
```diff
@@ -8,15 +8,16 @@
...
     "chalk": "^3.0.0",
     "cls-hooked": "^4.2.2",
+    "debug": "^4.1.1",
     "egg": "^2.26.1",
     "fs-extra": "^8.1.0",
     "glob": "^7.1.6",
     "glob-promise": "^3.4.0",
     "jimp": "^0.10.0",
     "midway": "^1.19.0",
+    "request-promise": "^4.2.5",
```
Comment (yorkie) [inline packages/daemon/package.json:20]: Missing `@types/request-promise`.

Comment (FeelyChau) [reply packages/daemon/package.json:21]: Done.

File: .github/workflows/cli.yml [hunk 6]
```diff
@@ -57,9 +57,12 @@ jobs:
       run: |
         ./packages/cli/dist/bin/pipcook init
         ./packages/cli/dist/bin/pipcook daemon start
-    - name: pipcook run
+    - name: pipcook run local path
       run: |
         ./packages/cli/dist/bin/pipcook run ./test/pipelines/text-bayes-classification.json --verbose
+    - name: pipcook run url
```
Comment (yorkie) [inline .github/workflows/cli.yml:63]:
```suggestion
    - name: pipcook run from url
```

File: .github/workflows/cli.yml [hunk 7]
```diff
@@ -57,9 +57,12 @@ jobs:
       run: |
         ./packages/cli/dist/bin/pipcook init
         ./packages/cli/dist/bin/pipcook daemon start
-    - name: pipcook run
+    - name: pipcook run local path
```
Comment (yorkie) [inline .github/workflows/cli.yml:60]:
```suggestion
    - name: pipcook run from local filename
```

File: packages/daemon/src/runner/helper.ts (hunk 2 above)
Comment (yorkie) [reply packages/daemon/src/runner/helper.ts:18]: It acts as the pathname for pipeline config.

File: packages/cli/src/actions/start.ts [hunk 8]
```diff
@@ -15,10 +16,18 @@ const start: StartHandler = async (filename: string, opts: any) => {
...
-    spinner.fail(`${filename} not exists`);
-    return process.exit(1);
+  const urlObj = url.parse(filename);
+  if (urlObj.protocol) {
+    if (urlObj.protocol !== 'http:' && urlObj.protocol !== 'https:') {
+      spinner.fail(`protocol ${urlObj.protocol} is not supported`);
+      return process.exit(1);
+    }
+  } else {
+    filename = path.isAbsolute(filename) ? filename : path.join(process.cwd(), filename);
```
Comment (yorkie) [inline packages/cli/src/actions/start.ts:26]: We could convert it to the file-protocol.

File: packages/daemon/src/runner/helper.ts [hunk 9]
```diff
@@ -13,8 +15,19 @@ import {
...
-  const configJson: RunConfigI = await fs.readJson(configPath);
+export async function parseConfig(config: string, generateId = true): Promise<PipelineDB> {
+  let configJson: RunConfigI = null;
+  let urlObj = null;
+  urlObj = url.parse(config);
+  if (urlObj.protocol) {
+    if (urlObj.protocol === 'http:' || urlObj.protocol === 'https:') {
+      configJson = JSON.parse(await request(config));
+    } else {
+      throw new Error(`protocol ${urlObj.protocol} is not supported`);
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:26]:
```suggestion
      throw new TypeError(`protocol ${urlObj.protocol} is not supported`);
```

File: packages/cli/src/actions/start.ts (hunk 8 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts:26]: How about we refactoring this into file-upload.

Comment (yorkie) [reply packages/cli/src/actions/start.ts:26]: You added the extra check at daemon, so I think it's required at this PR :)

**Comment by FeelyChau** [comment]
> Besides, IMO we can add a new step to run a pipeline remotely in our [`workflows/cli.yml`](https://github.com/alibaba/pipcook/blob/master/.github/workflows/cli.yml#L62) :)

Done.

File: packages/cli/src/actions/start.ts [hunk 10]
```diff
@@ -15,10 +16,20 @@ const start: StartHandler = async (filename: string, opts: any) => {
...
+    if (!existsSync(filename)) {
+      spinner.fail(`${filename} not exists`);
+      return process.exit(1);
+    }
+    urlObj = url.parse(`file://${filename}`);
+    filename = urlObj.href;
```
Comment (yorkie) [inline packages/cli/src/actions/start.ts:32]:
```js
const urlObj = url.parse(filename);
if (urlObj.protocol == null) {
  // FIXME(feely): file by default if protocol is null
  filename = path.isAbsolute(filename) ? filename : path.join(process.cwd(), filename);
  // check the filename existence
  if (!existsSync(filename)) {
    return process.exit(1);
  }
  filename = `file://${filename}`;
} else if (urlObj.protocol.indexOf([ 'http:', 'https:' ]) === -1) {
  return process.exit(1);
}
```

File: packages/daemon/src/runner/helper.ts [hunk 11]
```diff
@@ -13,8 +15,21 @@ import {
 
 const { PLUGINS, PIPCOOK_LOGS } = constants;
 
-export async function parseConfig(configPath: string, generateId = true): Promise<PipelineDB> {
-  const configJson: RunConfigI = await fs.readJson(configPath);
+export async function parseConfig(config: string, generateId = true): Promise<PipelineDB> {
+  let configJson: RunConfigI = null;
+  let urlObj = null;
+  urlObj = url.parse(config);
+  if (urlObj.protocol) {
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:22]: We can throw it firstly just like:

```js
if (urlObj.protocol == null) {
  throw TypeError(...);
}
```

File: run_pipeline.sh [hunk 12]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env sh
 
-./packages/cli/dist/bin/pipcook run ./test/pipelines/$1.json --verbose
+./packages/cli/dist/bin/pipcook run "./test/pipelines/$1.json" --verbose
```
Comment (yorkie) [inline run_pipeline.sh:3]: Which case do you make this change?

File: packages/daemon/src/runner/helper.ts [hunk 13]
```diff
@@ -14,7 +16,20 @@ import {
 const { PLUGINS, PIPCOOK_LOGS } = constants;
 
 export async function parseConfig(configPath: string, generateId = true): Promise<PipelineDB> {
-  const configJson: RunConfigI = await fs.readJson(configPath);
+  let configJson: RunConfigI = null;
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:19]:
```suggestion
  let configJson: RunConfigI;
```

File: packages/daemon/src/runner/helper.ts (hunk 2 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts]: OK.

File: packages/daemon/src/runner/helper.ts [hunk 14]
```diff
@@ -14,7 +16,20 @@ import {
...
+  let urlObj = null;
```
Comment (yorkie) [inline packages/daemon/src/runner/helper.ts:20]:
```suggestion
  let urlObj = url.parse(configPath);
```

File: run_pipeline.sh (hunk 12 above)
Comment (FeelyChau) [reply run_pipeline.sh:3]: Filename with space.

File: packages/cli/src/actions/start.ts (hunk 10 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts:32]: `filename` should be encoded if we use file-protocol.

File: packages/daemon/src/runner/helper.ts (hunk 14 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts]: Done.

File: packages/cli/src/actions/start.ts (hunk 10 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts]: Done.

File: packages/daemon/src/runner/helper.ts (hunk 11 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts]: Done.

File: packages/daemon/src/runner/helper.ts (hunk 13 above)
Comment (FeelyChau) [reply packages/daemon/src/runner/helper.ts]: Done.

File: packages/cli/src/actions/start.ts [hunk 15]
```diff
@@ -15,9 +16,19 @@ const start: StartHandler = async (filename: string, opts: any) => {
...
+  let urlObj = url.parse(filename);
+  // file default if the protocol is null
+  if (urlObj.protocol === null) {
+    filename = path.isAbsolute(filename) ? filename : path.join(process.cwd(), filename);
+    // check the filename existence
+    if (!existsSync(filename)) {
+      spinner.fail(`${filename} not exists`);
+      return process.exit(1);
+    } else {
+      filename = url.parse(`file://${filename}`).href;
```
Comment (yorkie) [inline packages/cli/src/actions/start.ts:28]: IMO, why not using `file://${filename}` directly?

File: packages/daemon/.github/workflows/nodejs.yml [hunk 16]
```diff
@@ -0,0 +1,31 @@
+# This workflow will do a clean install of node dependencies, build the source code and run tests across different versions of node
```
Comment (yorkie) [inline packages/daemon/.github/workflows/nodejs.yml:1]: What's this new file for?

File: run_pipeline.sh (hunk 12 above)
Comment (yorkie) [reply run_pipeline.sh:3]: Do we have pipeline filename with spaces?

Comment (FeelyChau) [reply run_pipeline.sh:3]: It's depended on the user.

File: packages/cli/src/actions/start.ts (hunk 15 above)
Comment (FeelyChau) [reply packages/cli/src/actions/start.ts:28]: The string `file://${filename}` could be an unsafe URI.
Look at:
https://tools.ietf.org/html/rfc3986#section-2

File: packages/daemon/.github/workflows/nodejs.yml (hunk 16 above)
Comment (FeelyChau) [reply packages/daemon/.github/workflows/nodejs.yml]: Unused file, removed.

**Comment by utkobe** [comment]
LGTM :)

**Pull request merged**